File: rd-net/RdFramework/SerializationCtx.cs
using System.Collections.Generic;
using JetBrains.Annotations;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Serialization;

#nullable disable

namespace JetBrains.Rd
{
  public struct SerializationCtx
  {
    public ISerializers Serializers { get; private set; }
    public IIdentities Identities { get; }

    [NotNull] public readonly IDictionary<string, IInternRoot<object>> InternRoots;

    public SerializationCtx(ISerializers serializers, IIdentities identities, IDictionary<string, IInternRoot<object>> internRoots = null) : this()
    {
      Serializers = serializers;
      Identities = identities;
      InternRoots = internRoots ?? new Dictionary<string, IInternRoot<object>>();
    }

    public SerializationCtx(IProtocol protocol, IDictionary<string, IInternRoot<object>> internRoots = null) : this(protocol.Serializers, protocol.Identities, internRoots)
    {
    }

    public SerializationCtx WithInternRootsHere(RdBindableBase owner, params string[] keys)
    {
      var newInternRoots = new Dictionary<string, IInternRoot<object>>(InternRoots);
      var identities = Identities;
      foreach (var key in keys)
      {
        var root = owner.GetOrCreateHighPriorityExtension("InternRoot-" + key, () => new InternRoot<object> { RdId = identities.Mix(owner.RdId, ".InternRoot-" + key) });
        newInternRoots[key] = root;
      }
      return new SerializationCtx(Serializers, Identities, newInternRoots);
    }
    
    public T ReadInterned<T>(UnsafeReader stream, string internKey, CtxReadDelegate<T> readValueDelegate)
    {
      if (!InternRoots.TryGetValue(internKey, out var interningRoot))
        return readValueDelegate(this, stream);

      var id = InternId.Read(stream);
      if (id.IsValid)
        return interningRoot.UnIntern<T>(id);
      return readValueDelegate(this, stream);
    }

    public void WriteInterned<T>(UnsafeWriter stream, T value, string internKey, CtxWriteDelegate<T> writeValueDelegate)
    {
      if (!InternRoots.TryGetValue(internKey, out var interningRoot))
      {
        writeValueDelegate(this, stream, value);
        return;
      }

      var id = interningRoot.Intern(value);
      InternId.Write(stream, id);
      if (!id.IsValid)
        writeValueDelegate(this, stream, value);
    }
  }
}

File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/SerializationCtx.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.framework.base.RdBindableBase
import com.jetbrains.rd.framework.impl.InternRoot
import com.jetbrains.rd.framework.impl.readInternId
import com.jetbrains.rd.framework.impl.writeInternId

class SerializationCtx(val serializers: ISerializers, val internRoots: Map<String, IInternRoot<Any>> = emptyMap()) {
    constructor(protocol: IProtocol) : this(protocol.serializers)
}

object Polymorphic : ISerializer<Any?> {
    @Suppress("UNCHECKED_CAST")
    operator fun <T> invoke() : ISerializer<T> = this as ISerializer<T>

    override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): Any? =
        ctx.serializers.readPolymorphicNullable(ctx, buffer)

    override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: Any?) =
        ctx.serializers.writePolymorphicNullable(ctx, buffer, value)
}

@Suppress("UNCHECKED_CAST")
class AbstractPolymorphic<T>(val declaration: IAbstractDeclaration<T>) : ISerializer<T> {
    @Suppress("UNCHECKED_CAST")
    operator fun <T> invoke() : ISerializer<T> = this as ISerializer<T>

    override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): T =
        ctx.serializers.readPolymorphicNullable(ctx, buffer, declaration) as T

    override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: T) =
        ctx.serializers.writePolymorphicNullable(ctx, buffer, value)
}

fun <T> ISerializer<T>.list() : ISerializer<List<T>> = object : ISerializer<List<T>> {
    override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): List<T> = buffer.readList { this@list.read(ctx, buffer) }
    override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: List<T>) = buffer.writeList(value) { this@list.write(ctx, buffer, it)}
}

inline fun <reified T> ISerializer<T>.array() : ISerializer<Array<T>> = object : ISerializer<Array<T>> {
    override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): Array<T> = buffer.readArray { this@array.read(ctx, buffer) }
    override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: Array<T>) = buffer.writeArray(value) { this@array.write(ctx, buffer, it)}
}

fun <T : Any> ISerializer<T>.nullable() : ISerializer<T?> = object : ISerializer<T?> {
    override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): T? = buffer.readNullable { this@nullable.read(ctx, buffer) }
    override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: T?) = buffer.writeNullable(value) { this@nullable.write(ctx, buffer, it)}
}

fun <T: Any> ISerializer<T>.interned(internKey: String) : ISerializer<T> = object : ISerializer<T> {
    override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): T = ctx.readInterned(buffer, internKey, this@interned::read)
    override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: T) = ctx.writeInterned(buffer, value, internKey, this@interned::write)
}

fun SerializationCtx.withInternRootsHere(owner: RdBindableBase, vararg newRoots: String): SerializationCtx {
    return SerializationCtx(serializers, internRoots.plus(newRoots.associate {
        it to owner.getOrCreateHighPriorityExtension("InternRoot-$it") { InternRoot<Any>() }.apply { rdid = owner.rdid.mix(".InternRoot-$it") } }))
}

inline fun <T: Any> SerializationCtx.readInterned(stream: AbstractBuffer, internKey: String, readValueDelegate: (SerializationCtx, AbstractBuffer) -> T): T {
    val internRoot = internRoots[internKey] ?: return readValueDelegate(this, stream)
    val internId = stream.readInternId()
    return if (internId.isValid)
        internRoot.unIntern(internId)
    else
        readValueDelegate(this, stream)
}

inline fun <T: Any> SerializationCtx.writeInterned(stream: AbstractBuffer, value: T, internKey: String, writeValueDelegate: (SerializationCtx, AbstractBuffer, T) -> Unit) {
    val internRoot = internRoots[internKey] ?: return writeValueDelegate(this, stream, value)
    val internId = internRoot.intern(value)
    stream.writeInternId(internId)
    if (!internId.isValid) // value couldn't be interned, send as-is
        writeValueDelegate(this, stream, value)
}